File: typescript/tests/easy/issue-templates-trials.test.ts
/**
 * Issuer.issue() — template inheritance + trial-issuance behaviour.
 *
 * Covers:
 *   - Issue from a flat template (defaults inherited).
 *   - Issue from a child template (parent merge wins on missing keys).
 *   - Per-call override beats template default.
 *   - Trial issuance recorded; same-fingerprint within cooldown rejected.
 *   - Trial issuance allowed after cooldown elapses.
 *   - Trial without pepper / fingerprint surfaces a clear error.
 */

import { describe, expect, it } from 'bun:test';

import { Licensing } from '@anorebel/licensing';
import type { LicenseTemplateInput } from '@anorebel/licensing/storage';
import { MemoryStorage } from '@anorebel/licensing/storage/memory';
import { hashFingerprint } from '../../src/trials/pepper.ts';

const PASSPHRASE = 'test-passphrase-must-be-at-least-32-chars';
const PEPPER = 'a'.repeat(32);

function tmplInput(overrides: Partial<LicenseTemplateInput> = {}): LicenseTemplateInput {
  return {
    scope_id: null,
    parent_id: null,
    name: overrides.name ?? `tmpl-${Math.random().toString(36).slice(2, 8)}`,
    max_usages: 5,
    trial_duration_sec: 0,
    trial_cooldown_sec: null,
    grace_duration_sec: 0,
    force_online_after_sec: null,
    entitlements: {},
    meta: {},
    ...overrides,
  };
}

describe('Issuer.issue() — template inheritance', () => {
  it('inherits maxUsages from a flat template', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({ db, signing: { passphrase: PASSPHRASE } });
    const template = await db.createTemplate(
      tmplInput({ name: 'pro', max_usages: 10, entitlements: { tier: 'pro' } }),
    );
    const license = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_1',
      templateId: template.id,
    });
    expect(license.raw.max_usages).toBe(10);
    expect(license.raw.template_id).toBe(template.id);
    // Resolved entitlements landed on the license meta.
    expect(license.raw.meta.entitlements).toEqual({ tier: 'pro' });
  });

  it('per-call maxUsages overrides template default', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({ db, signing: { passphrase: PASSPHRASE } });
    const template = await db.createTemplate(tmplInput({ name: 'pro', max_usages: 10 }));
    const license = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_2',
      templateId: template.id,
      maxUsages: 50,
    });
    expect(license.raw.max_usages).toBe(50);
  });

  it('child template merges entitlements with parent (child wins on conflict)', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({ db, signing: { passphrase: PASSPHRASE } });
    const parent = await db.createTemplate(
      tmplInput({ name: 'parent', max_usages: 5, entitlements: { tier: 'basic', seats: 5 } }),
    );
    const child = await db.createTemplate(
      tmplInput({
        name: 'child',
        parent_id: parent.id,
        max_usages: 10,
        entitlements: { tier: 'pro' },
      }),
    );
    const license = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_3',
      templateId: child.id,
    });
    // tier from child, seats from parent.
    expect(license.raw.meta.entitlements).toEqual({ tier: 'pro', seats: 5 });
    // Inheritable scalar (max_usages) from the leaf, not the merged ancestor.
    expect(license.raw.max_usages).toBe(10);
  });

  it('issue() requires maxUsages when no template is supplied', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({ db, signing: { passphrase: PASSPHRASE } });
    await expect(
      issuer.issue({
        licensableType: 'User',
        licensableId: 'u_4',
        // no templateId, no maxUsages
      }),
    ).rejects.toThrow(/requires `maxUsages`/);
  });
});

describe('Issuer.issue() — trial issuance', () => {
  const FINGERPRINT = 'fingerprint-canonical-input';

  it('records a trial dedupe row + sets is_trial flag in meta', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({
      db,
      signing: { passphrase: PASSPHRASE },
      trialPepper: PEPPER,
    });
    const template = await db.createTemplate(
      tmplInput({ name: 'trial', trial_duration_sec: 14 * 86400, trial_cooldown_sec: 86400 }),
    );
    const license = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_t',
      templateId: template.id,
      isTrial: true,
      fingerprint: FINGERPRINT,
    });
    expect(license.raw.meta.is_trial).toBe(true);
    // expiresAt derived from template.trial_duration_sec (14 days).
    expect(license.expiresAt).not.toBeNull();

    // The dedupe row must exist under the hash the issuer actually wrote.
    // Deriving it with the same keyed hash the issuer uses is what makes
    // this an assertion about persistence rather than about a constant:
    // a hardcoded literal here would simply return null forever and the
    // test would still pass.
    const expectedHash = hashFingerprint(PEPPER, FINGERPRINT);
    const found = await db.findTrialIssuance({
      template_id: template.id,
      fingerprint_hash: expectedHash,
    });
    expect(found).not.toBeNull();
    expect(found?.template_id).toBe(template.id);
    expect(found?.fingerprint_hash).toBe(expectedHash);

    // The raw fingerprint must never be persisted — only its keyed hash.
    expect(expectedHash).not.toContain(FINGERPRINT);
    expect(JSON.stringify(found)).not.toContain(FINGERPRINT);
  });

  it('rejects re-trial within the cooldown window', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({
      db,
      signing: { passphrase: PASSPHRASE },
      trialPepper: PEPPER,
    });
    const template = await db.createTemplate(
      tmplInput({ name: 'trial', trial_duration_sec: 86400, trial_cooldown_sec: 86400 }),
    );
    await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_a',
      templateId: template.id,
      isTrial: true,
      fingerprint: FINGERPRINT,
    });
    await expect(
      issuer.issue({
        licensableType: 'User',
        licensableId: 'u_a2',
        templateId: template.id,
        isTrial: true,
        fingerprint: FINGERPRINT,
      }),
    ).rejects.toMatchObject({ code: 'TrialAlreadyIssued' });
  });

  it('different fingerprints under the same template both succeed', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({
      db,
      signing: { passphrase: PASSPHRASE },
      trialPepper: PEPPER,
    });
    const template = await db.createTemplate(
      tmplInput({ name: 'trial', trial_duration_sec: 86400, trial_cooldown_sec: 86400 }),
    );
    const a = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_x',
      templateId: template.id,
      isTrial: true,
      fingerprint: 'fp-a',
    });
    const b = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_y',
      templateId: template.id,
      isTrial: true,
      fingerprint: 'fp-b',
    });
    expect(a.id).not.toBe(b.id);
  });

  it('different templates with the same fingerprint both succeed', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({
      db,
      signing: { passphrase: PASSPHRASE },
      trialPepper: PEPPER,
    });
    const t1 = await db.createTemplate(tmplInput({ name: 't1', trial_duration_sec: 86400 }));
    const t2 = await db.createTemplate(tmplInput({ name: 't2', trial_duration_sec: 86400 }));
    await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_p',
      templateId: t1.id,
      isTrial: true,
      fingerprint: FINGERPRINT,
    });
    const second = await issuer.issue({
      licensableType: 'User',
      licensableId: 'u_q',
      templateId: t2.id,
      isTrial: true,
      fingerprint: FINGERPRINT,
    });
    expect(second.id).toBeDefined();
  });

  it('trial without `trialPepper` config errors with a clear message', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({ db, signing: { passphrase: PASSPHRASE } });
    await expect(
      issuer.issue({
        licensableType: 'User',
        licensableId: 'u_n',
        maxUsages: 1,
        isTrial: true,
        fingerprint: FINGERPRINT,
      }),
    ).rejects.toThrow(/trialPepper/);
  });

  it('trial without `fingerprint` errors', async () => {
    const db = new MemoryStorage();
    const issuer = await Licensing.issuer({
      db,
      signing: { passphrase: PASSPHRASE },
      trialPepper: PEPPER,
    });
    await expect(
      issuer.issue({
        licensableType: 'User',
        licensableId: 'u_n',
        maxUsages: 1,
        isTrial: true,
      }),
    ).rejects.toThrow(/requires `fingerprint`/);
  });
});

File: licensing/easy/issue_templates_trials_test.go
package easy_test

import (
	"context"
	"errors"
	"strings"
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	"github.com/AnoRebel/licensing/licensing/easy"
	"github.com/AnoRebel/licensing/licensing/storage/memory"
)

const pepper = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" // 32 chars

func mkTemplate(t *testing.T, db lic.Storage, name string, mut func(*lic.LicenseTemplateInput)) *lic.LicenseTemplate {
	t.Helper()
	in := lic.LicenseTemplateInput{
		Name:             name,
		MaxUsages:        5,
		TrialDurationSec: 0,
		GraceDurationSec: 0,
		Entitlements:     map[string]any{},
		Meta:             map[string]any{},
	}
	if mut != nil {
		mut(&in)
	}
	tmpl, err := db.CreateTemplate(in)
	if err != nil {
		t.Fatalf("create template: %v", err)
	}
	return tmpl
}

func TestIssue_InheritsMaxUsagesFromTemplate(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:      db,
		Signing: &easy.SigningConfig{Passphrase: passphrase},
	})
	if err != nil {
		t.Fatal(err)
	}
	tmpl := mkTemplate(t, db, "pro", func(in *lic.LicenseTemplateInput) {
		in.MaxUsages = 10
		in.Entitlements = map[string]any{"tier": "pro"}
	})
	license, err := issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_1",
		TemplateID:     &tmpl.ID,
	})
	if err != nil {
		t.Fatal(err)
	}
	if license.MaxUsages != 10 {
		t.Fatalf("max_usages: want 10, got %d", license.MaxUsages)
	}
	ent, ok := license.Meta["entitlements"].(map[string]any)
	if !ok || ent["tier"] != "pro" {
		t.Fatalf("entitlements not propagated: %v", license.Meta["entitlements"])
	}
}

func TestIssue_PerCallMaxUsagesOverridesTemplate(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:      db,
		Signing: &easy.SigningConfig{Passphrase: passphrase},
	})
	if err != nil {
		t.Fatal(err)
	}
	tmpl := mkTemplate(t, db, "pro", func(in *lic.LicenseTemplateInput) { in.MaxUsages = 10 })
	license, err := issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_2",
		TemplateID:     &tmpl.ID,
		MaxUsages:      50,
		MaxUsagesSet:   true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if license.MaxUsages != 50 {
		t.Fatalf("override failed: got %d", license.MaxUsages)
	}
}

func TestIssue_ChildTemplateMergesEntitlements(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:      db,
		Signing: &easy.SigningConfig{Passphrase: passphrase},
	})
	if err != nil {
		t.Fatal(err)
	}
	parent := mkTemplate(t, db, "parent", func(in *lic.LicenseTemplateInput) {
		in.Entitlements = map[string]any{"tier": "basic", "seats": 5}
	})
	child := mkTemplate(t, db, "child", func(in *lic.LicenseTemplateInput) {
		in.ParentID = &parent.ID
		in.MaxUsages = 10
		in.Entitlements = map[string]any{"tier": "pro"}
	})
	license, err := issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_3",
		TemplateID:     &child.ID,
	})
	if err != nil {
		t.Fatal(err)
	}
	ent := license.Meta["entitlements"].(map[string]any)
	// Note: the resolver merges into a typed map but our scan reads back as
	// generic map[string]any after JSON round-trip; numeric becomes float64.
	if ent["tier"] != "pro" {
		t.Fatalf("tier: want pro, got %v", ent["tier"])
	}
	// seats survives from parent.
	if ent["seats"] == nil {
		t.Fatalf("seats not inherited from parent: %v", ent)
	}
}

func TestIssue_RequiresMaxUsagesWithoutTemplate(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:      db,
		Signing: &easy.SigningConfig{Passphrase: passphrase},
	})
	if err != nil {
		t.Fatal(err)
	}
	_, err = issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_4",
		// no TemplateID, no MaxUsages
	})
	if err == nil || !strings.Contains(err.Error(), "MaxUsages") {
		t.Fatalf("expected MaxUsages error, got %v", err)
	}
}

func TestIssue_TrialDedupe(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:          db,
		Signing:     &easy.SigningConfig{Passphrase: passphrase},
		TrialPepper: pepper,
	})
	if err != nil {
		t.Fatal(err)
	}
	cooldown := 86400
	tmpl := mkTemplate(t, db, "trial", func(in *lic.LicenseTemplateInput) {
		in.TrialDurationSec = 86400
		in.TrialCooldownSec = &cooldown
	})
	_, err = issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_a",
		TemplateID:     &tmpl.ID,
		IsTrial:        true,
		Fingerprint:    "fp-1",
	})
	if err != nil {
		t.Fatal(err)
	}
	_, err = issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_a2",
		TemplateID:     &tmpl.ID,
		IsTrial:        true,
		Fingerprint:    "fp-1",
	})
	var lerr *lic.Error
	if !errors.As(err, &lerr) || lerr.Code != lic.CodeTrialAlreadyIssued {
		t.Fatalf("expected CodeTrialAlreadyIssued, got %v", err)
	}
}

func TestIssue_TrialDifferentFingerprintsBothSucceed(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:          db,
		Signing:     &easy.SigningConfig{Passphrase: passphrase},
		TrialPepper: pepper,
	})
	if err != nil {
		t.Fatal(err)
	}
	tmpl := mkTemplate(t, db, "trial", func(in *lic.LicenseTemplateInput) {
		in.TrialDurationSec = 86400
	})
	if _, err := issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User", LicensableID: "u_x", TemplateID: &tmpl.ID,
		IsTrial: true, Fingerprint: "fp-a",
	}); err != nil {
		t.Fatal(err)
	}
	if _, err := issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User", LicensableID: "u_y", TemplateID: &tmpl.ID,
		IsTrial: true, Fingerprint: "fp-b",
	}); err != nil {
		t.Fatal(err)
	}
}

func TestIssue_TrialWithoutPepperErrors(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:      db,
		Signing: &easy.SigningConfig{Passphrase: passphrase},
		// No TrialPepper.
	})
	if err != nil {
		t.Fatal(err)
	}
	_, err = issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_n",
		MaxUsages:      1,
		MaxUsagesSet:   true,
		IsTrial:        true,
		Fingerprint:    "fp",
	})
	if err == nil || !strings.Contains(err.Error(), "TrialPepper") {
		t.Fatalf("expected TrialPepper error, got %v", err)
	}
}

func TestIssue_TrialWithoutFingerprintErrors(t *testing.T) {
	db := memory.New(memory.Options{})
	issuer, err := easy.NewIssuer(easy.IssuerConfig{
		DB:          db,
		Signing:     &easy.SigningConfig{Passphrase: passphrase},
		TrialPepper: pepper,
	})
	if err != nil {
		t.Fatal(err)
	}
	_, err = issuer.Issue(context.Background(), easy.IssueInput{
		LicensableType: "User",
		LicensableID:   "u_n",
		MaxUsages:      1,
		MaxUsagesSet:   true,
		IsTrial:        true,
		// no Fingerprint
	})
	if err == nil || !strings.Contains(err.Error(), "Fingerprint") {
		t.Fatalf("expected Fingerprint error, got %v", err)
	}
}
